### PR TITLE
DAOS-4420 dfuse: Reimplement readdir() for performance.

### PR DESCRIPTION
**Pull request by ashleypittman**
Redo the readdir implementation to be able to support both readdirplus
and adaptive readdir.  Rather than packing the data into buffers
between calls save just the dentry names, then on subsequent calls
use the correct packing function.  This allows dfuse to reply to either
type of request on any directory stream, and that should lead to
performance improvements for many common use cases.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

### CODE REVIEW

**Review by jolivier23 — DISMISSED** [review]

**Review by mchaarawi — CHANGES_REQUESTED** [review]
before reviewing this, i ran my small program that i had when adding the readdir implementation that tests
1) seekdir / telldir functionality
2) does a readdir over all entries in a dir with a large number of entries.

seekdir/telldir is broken as a result of this patch.
performance of readdir is worse.

we definitely can't land this until 1 is fixed, and if you can explain in what common cases performance improvement is expected here since i see the opposite in the basic readdir case?

**Comment by ashleypittman** [comment]
> before reviewing this, i ran my small program that i had when adding the readdir implementation that tests
> 
> 1. seekdir / telldir functionality
> 2. does a readdir over all entries in a dir with a large number of entries.
> 
> seekdir/telldir is broken as a result of this patch.
> performance of readdir is worse.
> 
> we definitely can't land this until 1 is fixed, and if you can explain in what common cases performance improvement is expected here since i see the opposite in the basic readdir case?

Ok, I can try writing some tests for it tomorrow for 1.  Were you doing seekdir to 0 or to some intermediate point?

The advantage of this is adaptive readdirplus, so any command doing readdir + stat should be faster, the adaptive nature means that it'll do readdirplus for the first call, and if the application doesn't call stat then it'll do plain readdir for subsequent ones, so the first readdir call for will be slower if your test isn't also calling stat.

Some tuning of buffer sizes (or rather entries) to return would help, although the number here actually depends on the length of the filename so it's not easy to tune for the general case.

**Comment by mchaarawi** [comment]
> Ok, I can try writing some tests for it tomorrow for 1. Were you doing seekdir to 0 or to some intermediate point?

my program was doing something like:

`        if ((dir1 = opendir(argv[1])) == NULL)
                perror("opendir() error");
#if 1
        if ((dir2 = opendir(argv[1])) == NULL)
                perror("opendir() error");

        printf("contents of %s:\n", argv[1]);

        printf("Position without read: %lld\n", telldir(dir1));

        entry1 = readdir(dir1);
        printf("START: %s\n", entry1->d_name);
        off[0] = telldir(dir1);                                                                                                                                                                                 
        printf("Position after 1 read: %lld\n", telldir(dir1));

        entry1 = readdir(dir1);
        printf("NEXT: %s\n", entry1->d_name);
        off[1] = telldir(dir1);
        printf("Position after 2 read: %lld\n", telldir(dir1));

        entry1 = readdir(dir1);
        printf("NEXT: %s\n", entry1->d_name);
        off[2] = telldir(dir1);
        printf("Position after 3 read: %lld\n", telldir(dir1));`

and I expect to see something like:

`Position without read: 0
START: d7
Position after 1 read: 1
NEXT: d3
Position after 2 read: 2
NEXT: d5
Position after 3 read: 3
`

with your patch i see:

`Position without read: 0
START: d7
Position after 1 read: 2
NEXT: d3
Position after 2 read: 3
NEXT: d5
Position after 3 read: 4`

**Comment by ashleypittman** [comment]
Merged with master to resolve conflicts.

**Review by jolivier23 — DISMISSED** [review]

**Review by mchaarawi — CHANGES_REQUESTED** [review]
my seekdir/telldir verification still fails. maybe you should add some tests to verify.
also readdir performance is still about 50% worse with my quick testing.

**Comment by ashleypittman** [comment]
> my seekdir/telldir verification still fails. maybe you should add some tests to verify.
> also readdir performance is still about 50% worse with my quick testing.

The performance is massively dictated by the hash table in dfuse, I make this PR about 3% slower in my testing, but once this is landed and I can work on the HT changes there's a 80% improvement.  I've written a little readdir performance test which I'll send as a different PR, but the results are:

```
master
        10 create_time 0.03 dirs 0.00 files 0.00 dirs_with_stat 0.01 files_with_stat 0.01
       100 create_time 0.00 dirs 0.01 files 0.01 dirs_with_stat 0.05 files_with_stat 0.07
      1000 create_time 0.00 dirs 0.08 files 0.09 dirs_with_stat 0.49 files_with_stat 0.68
     10000 create_time 0.00 dirs 0.91 files 1.16 dirs_with_stat 5.84 files_with_stat 8.77
    100000 create_time 0.00 dirs 20.58 files 44.33 dirs_with_stat 156.20 files_with_stat 283.30

this PR

        10 create_time 0.03 dirs 0.00 files 0.01 dirs_with_stat 0.01 files_with_stat 0.01
       100 create_time 0.00 dirs 0.01 files 0.01 dirs_with_stat 0.05 files_with_stat 0.07
      1000 create_time 0.00 dirs 0.08 files 0.09 dirs_with_stat 0.49 files_with_stat 0.70
     10000 create_time 0.00 dirs 0.88 files 1.16 dirs_with_stat 5.98 files_with_stat 9.17
    100000 create_time 0.00 dirs 21.18 files 45.90 dirs_with_stat 162.90 files_with_stat 301.09

master with HT

        10 create_time 0.03 dirs 0.01 files 0.00 dirs_with_stat 0.01 files_with_stat 0.01
       100 create_time 0.00 dirs 0.01 files 0.01 dirs_with_stat 0.05 files_with_stat 0.07
      1000 create_time 0.00 dirs 0.08 files 0.08 dirs_with_stat 0.48 files_with_stat 0.66
     10000 create_time 0.00 dirs 0.76 files 0.79 dirs_with_stat 4.94 files_with_stat 6.60
    100000 create_time 0.00 dirs 7.84 files 8.39 dirs_with_stat 48.71 files_with_stat 68.82

this PR with HT

        10 create_time 0.03 dirs 0.00 files 0.01 dirs_with_stat 0.01 files_with_stat 0.01
       100 create_time 0.00 dirs 0.01 files 0.01 dirs_with_stat 0.05 files_with_stat 0.07
      1000 create_time 0.00 dirs 0.08 files 0.08 dirs_with_stat 0.48 files_with_stat 0.69
     10000 create_time 0.00 dirs 0.74 files 0.79 dirs_with_stat 4.76 files_with_stat 6.84
    100000 create_time 0.00 dirs 7.62 files 8.28 dirs_with_stat 48.56 files_with_stat 70.15

```

As you can see this PR loses a couple of percentage, however I can make them back up later and it's certainly not the 50% you're seeing.

Interestingly I'm using 'ls -t' to sort by time, if I use 'ls -l' then ls will try and read a couple of acl extended attributes which significantly slows things down as well, I'd expect the HT changes to help here due to enabling LRU but I haven't specifically tested this.

**Comment by mchaarawi** [comment]
I can do a better perf evaluation to verify my findings. I didn't test many cases. but im ok with a few ms perf hit, but i was seeing more like 5 sec (~50%) which is no acceptable.
in any case, this can't land till you fix telldir/seekdir.
what i observe is that when you do some readdir() calls, then seekdir() back to an earlier position, and readdir() again, the first entry which you are supposed to enumerate at that seek seems to be skipped.

**Comment by ashleypittman** [comment]
I'm writing a test for seekdir()/telldir() currently.

**Comment by mchaarawi** [comment]
I still think there is something wrong here.
I tried more testing on boro with psm2 (1 server 1 client nodes):

Create dataset:
`mpirun -np 10 --hostfile ~/config/cli_hosts  mdtest -a DFS --dfs.cont $DAOS_CONT --dfs.svcl $DAOS_SVCL --dfs.pool $DAOS_POOL -n 10000 -d / --dfs.oclass=S1 -F -C`

Master
```
mschaara@boro-20:/tmp/dfuse/test-dir.0-0/mdtest_tree.0$ time /bin/ls . | wc -l
100000
real	0m16.422s
user	0m0.903s
sys	0m0.029s
mschaara@boro-20:/tmp/dfuse/test-dir.0-0/mdtest_tree.0$ time /bin/ls -l . | wc -l
100001
real	3m11.010s
user	0m1.583s
sys	0m3.200s
```

dfuse-readdirplus
```
mschaara@boro-20:/tmp/dfuse/test-dir.0-0/mdtest_tree.0$ time /bin/ls . | wc -l
101190
real	1m28.404s
user	0m0.916s
sys	0m0.021s
mschaara@boro-20:/tmp/dfuse/test-dir.0-0/mdtest_tree.0$ time /bin/ls -l . | wc -l
101887
real	5m18.579s
user	0m1.160s
sys	0m3.254s
```

you can see that aside from the performance, there is a wrong number of items reported by ls and ls -l with your branch.

**Review by jolivier23 — DISMISSED** [review]

**Comment by ashleypittman** [comment]
I tested the performance of this yesterday and it's similar to master in most cases but seems to be better when accessing files in readdirplus mode for some reason.  Today I wrote a test program that creates 5000 files, and then walks through the list forwards, iterates through files seeking forwards one at a time, and then starts at the end of the directory working backwards, seeking to a new location and reading until the end of directory and this passes in all cases.  There had previously been an issue where seek() would have worked but left the anchor in a incorrect state, so seek() then a succession of readdir() calls would eventually have returned incorrect entries.

**Comment by ashleypittman** [comment]
Rebase and whitespace changes for checkpatch only.

**Comment by ashleypittman** [comment]
Rebased because of conflicts.

**Comment by ashleypittman** [comment]
Bad merge last time so I've fixed it and rebased again.

**Comment by johannlombardi** [comment]
Need to be rebased.

**Review by jolivier23 — DISMISSED** [review]

**Review by mchaarawi — CHANGES_REQUESTED** [review]
@johannlombardi @ashleypittman
im not going to approve nor block this PR based on the performance claims.
But I have yet to see where the performance improvement is. In my testing i have not seen any performance improvement over the existing implementation and still do not see the value of landing this.

on the other hand, i will still block this PR due to a functional issue that i found that can be reproduced with the following steps:
- create container / mount dfuse (at /tmp/dfuse)
- cd /tmp/dfuse; mkdir test1
- mkdir test1/d{1..100000}
- un mount dfuse
- remount dfuse
- cd /tmp/dfuse; ls
- got something like:
mschaara@boro-20:/tmp/dfuse$ ls
ls: cannot access test1: Input/output error
test1

**Comment by ashleypittman** [comment]
> @johannlombardi @ashleypittman
> im not going to approve nor block this PR based on the performance claims.
> But I have yet to see where the performance improvement is. In my testing i have not seen any performance improvement over the existing implementation and still do not see the value of landing this.
> 
> on the other hand, i will still block this PR due to a functional issue that i found that can be reproduced with the following steps:
> 
> * create container / mount dfuse (at /tmp/dfuse)
> * cd /tmp/dfuse; mkdir test1
> * mkdir test1/d{1..100000}
> * un mount dfuse
> * remount dfuse
> * cd /tmp/dfuse; ls
> * got something like:
>   mschaara@boro-20:/tmp/dfuse$ ls
>   ls: cannot access test1: Input/output error
>   test1

I'll take a look.  I've been rebasing this PR and backporting some fixes but I haven't tested it as much as other branches.

https://github.com/daos-stack/daos/pull/3753 adds a test which happens to do exactly the above, could you review that please?

**Review by mchaarawi — APPROVED** [review]

**Comment by mchaarawi** [comment]
ran some tests with psm2.
functionally, this is finally working now in the test cases i had.
performance wise, i see no difference except a small gain in the case when caching is enabled with ls -la

File: src/client/dfuse/ops/lookup.c
Comment (daltonbohning) [inline src/client/dfuse/ops/lookup.c:195]: Minor format question.
I thought our preferred style was to NOT use a space here?
```suggestion
						  DF_RC"\n", DP_RC(rc));
```

File: src/client/dfuse/ops/readdir.c
Comment (daltonbohning) [inline src/client/dfuse/ops/readdir.c:97]: Only if you need to push again:
```suggestion
	d_list_t			*rlink;
```

**Review by daltonbohning — APPROVED** [review]
Functionally, this also seems to be working with the relevant datamover tools and use cases.

File: src/client/dfs/dfs.c
Comment (ashleypittman) [reply src/client/dfs/dfs.c:2409]: The change as a whole avoids a memory allocation and failure point, but this is just setting the memory back to how it was before.  This assignment probably isn't needed.

File: src/client/dfuse/ops/lookup.c
Comment (ashleypittman) [reply src/client/dfuse/ops/lookup.c:195]: You're entirely correct, this PR has been sitting a while.

File: src/client/dfuse/ops/readdir.c
Comment (ashleypittman) [reply src/client/dfuse/ops/readdir.c:97]: Will fix in a later PR.

**Pull request merged**
